### ISSUE TITLE
Feature request: use Gmail push notification API instead of polling

### ISSUE THREAD

**Issue by baip** [gauteh/gmailieer#84]
Thanks for the awesome tool. It makes Emacs + notmuch a serious mail client solution. It looks like [this API](https://developers.google.com/gmail/api/guides/push) lets you watch for changes to Gmail mailboxes. I'd like to propose implementing such push notification, which I imagine would be more preferable to regular polling.

**Comment by gauteh** [comment]
baip writes on June 8, 2018 4:14:
> Thanks for the awesome tool. It makes Emacs + notmuch a serious mail
> client solution. It looks like [this
> API](https://developers.google.com/gmail/api/guides/push) lets you
> watch for changes to Gmail mailboxes. I'd like to propose implementing
> such push notification, which I imagine would be more preferable to
> regular polling.

Hi, thanks. Yes - I think this would be pretty cool. It would require a
long-running gmailieer process, which could be invoked by something
like: `gmi watch`. The loaded cache and state has not been stress-tested for
long-running processes like this, but I think it is pretty consistent by now.

We have to decide how local-to-remote sync should be handled (e.g. check
at regular intervals for changes to notmuch db revision, as well as
check each time a change is received).

Additionally, the `gmi watch` command should take a command which is
called each time a change has been received so that a mail client can be
notified and refresh its view.

In short, I think this could be integrated into gmi as it is now with 
out much re-structuring! Did you plan to make an attempt at this 
yourself?

**Comment by gauteh** [comment]
Actually, when reading the API it seems that this can only be used to set up GMail to notify a server - but not notifying through a persistent HTTP connection. In which case this is unfortunately not useful for us. Am I reading this wrong?

**Comment by baip** [comment]
Yeah, it looks like that I jumped the gun suggesting this... I had assumed that IP addresses can be used as endpoints, but after reading the documentation more carefully, it does seem to require HTTPS and CA-signed SSL certificate, so not very useful indeed.

Another thought would be to use IMAP IDLE to watch selected mailboxes. This feels straight-forward to implement. I've been using [imapnotify](https://github.com/a-sk/node-imapnotify) in combination with gmailieer. There is also a [Python version](https://github.com/a-sk/python-imapnotify) that may be called directly or as implementation reference. I'd certainly be happy to play around and help wherever I can. For now, unfortunately, I don't think I'm familiar enough with gmailieer to do anything useful efficiently..

**Comment by gauteh** [comment]
Yeah, that might be a viable path. There’s a closed PR for opening IMAP
access (was trying out for performance reasons), but it requires much wider
access to the account.

lør. 9. jun. 2018 kl. 07:29 skrev baip <notifications@github.com>:

**Comment by gauteh** [comment]
Unless someone wants to try this I think I am going to close for now. 
Two main reasons:

* Using IMAP IDLE will require much greater access to the user account, 
  including write-access. We do not actually need to write something, 
  and I think we should try to avoid this if possible.

* Current (incremental) polling only retrieves the current historyId - 
  this is a very cheap request and it can probably be made very 
  frequently (seconds?) without problems. The rest of the handling would 
  be more or less the same. A simple loop with delay could therefore 
  probably be used to simulate RT fetching.

**Comment by comex** [comment]
FWIW, it *is* possible to wait for notifications on a persistent HTTP connection rather than using a webhook.  This is done with the [pull](https://cloud.google.com/pubsub/docs/reference/rest/v1/projects.subscriptions/pull) method.
